core/src/main/scala/kafka/log/Log.scala

```diff
@@ -2005,6 +2005,7 @@ class Log(@volatile var dir: File,
       lock synchronized {
         checkIfMemoryMappedBufferClosed()
         removeLogMetrics()
+        producerExpireCheck.cancel(true)
```

Comment:
It's a little surprising to me that we don't _always_ call `close()` on the `Log` object. Perhaps we can add a few comments which emphasize that the log is considered closed after deletion and no further cleanup is needed. Even better would be if we could consolidate these two paths somehow so that we do not have this kind of bug in the future.

Reply:
Maybe we can do the simple thing in this PR so that we can backport.

Reply:
I'm happy with that if improvements are not straightforward. Just whenever we punt these tech debts, they tend to never get done.

Reply:
Done.

Reply:
As discussed, limited this PR to only leak fix.